src/Java.Interop.Tools.Maven/Extensions/PropertyStack.cs

```diff
@@ -0,0 +1,42 @@
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Xml.Linq;
+using Java.Interop.Tools.Maven.Models;
+
+namespace Java.Interop.Tools.Maven.Extensions;
+
+class PropertyStack
+{
+	// Why go to this trouble?
+	// A property can be specified in both a child POM and its parent POM.
+	// Even if the property is being consumed in the parent POM, the property in
+	// the child POM takes precedence.
+	readonly List<List<KeyValuePair<string, string>>> stack = new ();
+
+	public void Push (ModelProperties? properties)
+	{
+		// We add a new list to the stack, even if it's empty, so that the Pop works later
+		var list = new List<KeyValuePair<string, string>> ();
+
+		if (properties?.Any is Collection<XElement> props)
+			foreach (var prop in props)
+				list.Add (new KeyValuePair<string, string> (prop.Name.LocalName, prop.Value));
+
+		stack.Add (list);
+	}
+
+	public void Pop ()
+	{
+		stack.RemoveAt (stack.Count - 1);
+	}
+
+	public string Apply (string value)
+	{
+		foreach (var property_set in stack) {
+			foreach (var prop in property_set)
+				value = value.Replace ($"${{{prop.Key}}}", prop.Value);
```

Comment:
This gives me *some* concern, if only because this *could* produce *lots* of garbage, all depending on `stack`.

I think I'd prefer a "optimization check" + use of [`StringBuilder.Replace()`](https://learn.microsoft.com/en-us/dotnet/api/system.text.stringbuilder.replace?view=net-8.0#system-text-stringbuilder-replace(system-string-system-string)):

```csharp
if (stack.Count == 0 || !value.Contains("${") {
    return value;
}
var s = new StringBuilder (value);
foreach (var property_set in stack) {
    foreach (var prop in property_set) {
        s.Replace ($"${{{prop.Key}}}", prop.Value);
    }
}
return s.ToString ();
```

Reply:
Added the optimization check.  👍 

From my research looking at POM files, this code will rarely be hit with the optimization check, and when it is hit it will generally be for small numbers of properties.  The general guidance seems to be that `string.Replace` is faster than `StringBuilder.Replace` for small numbers of replaces, despite the extra garbage, as it avoids the overhead of creating the `StringBuilder` and converting it `ToString` at the end.

Reply:
I don't recall reading the guidance about "`string.Replace` is faster … for small numbers of replace".  Would be interesting to read more.